pkg/provision/cleanup.go

```diff
@@ -0,0 +1,234 @@
+package provision
+
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/namespaces"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	"github.com/threefoldtech/tfexplorer/client"
+	"github.com/threefoldtech/tfexplorer/models/generated/workloads"
+	"github.com/threefoldtech/zbus"
+	"github.com/threefoldtech/zos/pkg"
+	"github.com/threefoldtech/zos/pkg/app"
+	"github.com/threefoldtech/zos/pkg/provision/common"
+	"github.com/threefoldtech/zos/pkg/storage"
+	"github.com/threefoldtech/zos/pkg/storage/filesystem"
+	"github.com/threefoldtech/zos/pkg/zdb"
+	"golang.org/x/net/context"
+)
+
+const (
+	containerdSock = "/run/containerd/containerd.sock"
+)
+
+// we declare this method as a variable so we can
+// mock it in testing.
+func initZdbConnection(id string) zdb.Client {
+	socket := fmt.Sprintf("unix://%s/zdb.sock", socketDir(id))
+	return zdb.New(socket)
+}
+
+func socketDir(containerID string) string {
+	return fmt.Sprintf("/var/run/zdb_%s", containerID)
+}
+
+// CleanupResources cleans up unused resources
+func CleanupResources(msgBrokerCon string) error {
+	client, err := app.ExplorerClient()
+	if err != nil {
+		return err
+	}
+
+	toSave, toDelete, err := checkContainers(msgBrokerCon)
+	if err != nil {
+		return errors.Wrap(err, "failed to check containers")
+	}
+
+	ctx := context.TODO()
+	utils := filesystem.NewUtils()
+	pools, err := ioutil.ReadDir("/mnt")
+	if err != nil {
+		return errors.Wrap(err, "failed to list available pools")
+	}
+
+	for _, pool := range pools {
+		if !pool.IsDir() {
+			continue
+		}
+		path := filepath.Join("/mnt", pool.Name())
+		subvols, err := utils.SubvolumeList(ctx, path)
+		if err != nil {
+			// A Pool might not be mounted here, skip it
+			log.Error().Err(err).Msgf("failed to list available subvolumes in: %s", path)
+			continue
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+			log.Info().Msgf("checking subvol %s", subvol.Path)
+			// Don't delete zos-cache!
+			if subvol.Path == storage.CacheLabel {
+				continue
+			}
+
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// Now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
```

Comment:
Why do we even need the toSave map? isn't it better to only fill a list of the `toDelete` and then just check if the volume is in that map or not ? I mean a volume can either be `to-save` or `to-delete` hence a one map is enough